### PR TITLE
Omit nools boilerplate for declarative configs > core 4.2

### LINKED ISSUE #225: compile-app-settings - Nools can't process scripts which `require('moment')`
1. Add `require('moment') to `tasks.js` or `targets.js`
1. Run `medic-conf compile-app-settings upload-app-settings`
1. Load webapp

**Expected** Glory
**Observed:** Crash

Invalid rule:
```
VM116315 inbox.js:53997 RuleEngine failed to Initialize Error: Invalid rule definition 
Unable to match { in {   when { c: Contact } then { 
```

When running with `-- --debug`, it gets a bit further: invalid action:
```
...
missing ) after argument list
```

### PR DESCRIPTION
**Pull request by kennsippell**
# Description

This is a companion change with https://github.com/medic/cht-core/pull/8065 proposed for cht-core 4.2.

Fixes #225 

Instead of always adding nools boilerplate to every compiled declarative configuration, this change adds it only when deploying to cht-core instances under 4.2.  Note the compiled app settings is augmented during upload instead of during compilation so `compile-app-settings` can remain tight, fast, and dependency-free for app developer productivity.

# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or integration tests where appropriate
- Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: src/lib/compile-nools-rules.js [hunk 1]
```diff
@@ -25,10 +25,14 @@ const compileNoolsRules = async (projectDir, options = {}) => {
       throw new Error(`Both legacy and declarative files found. You should either have rules.nools.js xor ${DECLARATIVE_NOOLS_FILES} files.`);
```
Comment (dianabarsan) [inline src/lib/compile-nools-rules.js:25]: Might I suggest we change this file name to exclude "nools" term? `compile-rules` ?

File: src/fn/upload-app-settings.js [hunk 2]
```diff
@@ -1,32 +1,53 @@
...
+// augmented during upload so compile-app-settings doesn't require a url on the command-line
+async function augmentDeclarativeWithNoolsBoilerplate(appSettings) {
+  if (!appSettings || !appSettings.tasks || !appSettings.tasks.isDeclarative) {
+    return;
+  }
+
+  const actualCoreVersion = getValidApiVersion(appSettings);
+  const addNoolsBoilerplate = semver.lt(actualCoreVersion, '4.2.0');
+  if (addNoolsBoilerplate) {
+    appSettings.tasks.rules =   `define Target { _id: null, contact: null, deleted: null, type: null, pass: null, date: null, groupBy: null }
```
Comment (dianabarsan) [inline src/fn/upload-app-settings.js:42]: This is an odd place to have this giant string. Can this maybe move to a library that will deal with old "nools" and have that library export a function that produces this result?

File: test/fn/upload-app-settings.spec.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+        }),
+        environment: {
+          pathToProject: ',',
+        },
+        fs: {
+          read: sinon.stub().returns(appSettings),
+        }
+      };
+
+      return uploadAppSettings.__with__(mocks)(async () => {
```
Comment (dianabarsan) [inline test/fn/upload-app-settings.spec.js:46]: I find `__with__` to be hard to read and follow with the callback style. 
I would much rather have:
```javascript
let uploadAppSettings;
...
beforeEach(() => {
  uploadAppSettings = rewire('../../src/fn/upload-app-settings');
});
...
uploadAppSettings.__set__(mocks);

await uploadAppSettings.execute();
expect(apiUpload.calledOnce).to.be.true;
const isNoolsAdded = apiUpload.args[0][0].tasks.rules !== rules;
expect(isNoolsAdded).to.eq(scenario.expectNools);
```

**Review by dianabarsan — APPROVED** [review]
I like this approach. All my comments are nits about code structure.

File: src/lib/compile-nools-rules.js (hunk 1 above)
Comment (kennsippell) [reply src/lib/compile-nools-rules.js:25]: I've renamed to `compile-rules.js`. As long as I'm moving files around - I've also moved all compilation related code into folders called `compilation` (one for src and one for test).

File: test/fn/upload-app-settings.spec.js (hunk 3 above)
Comment (kennsippell) [reply test/fn/upload-app-settings.spec.js:46]: legit

File: src/fn/upload-app-settings.js (hunk 2 above)
Comment (kennsippell) [reply src/fn/upload-app-settings.js:42]: Moved to a `nools-utils.js` file which also includes some weird nools code for minifying.

**Comment by kennsippell** [comment]
@dianabarsan [This](https://github.com/medic/cht-conf/pull/536#discussion_r1108329487) was a big enough change I'd welcome a second look.

**Review by dianabarsan — APPROVED** [review]
I enjoyed reading this! Thanks a lot for taking the time and making this refactor.

**Pull request merged**